### DISCUSSION TITLE
Serialization via 'toJSON` does not work correctly when loading shared relationships via a nested entity by QB

### DISCUSSION THREAD

**Discussion by mrmlnc** [mikro-orm/mikro-orm#6256] [General]
### Describe the bug

After upgrading from version 6.3.13 to version 6.4.0, the serialization behavior has changed.

Let's say we have an entity called `TestCase` and `TestCaseRevision`. Both of them have a relationship with the `TestCasePriority` entity. The entity `TestCaseRevision` has a relationship with `TestCase`.

If you request `TestCasePriority` via `TestCase` by building a QB query via `TestCaseRevision`, then the `priority` field of the test case revision will not be serialized during serialization by the `toJSON` method call. However, serialization will work correctly when the entity serialized via the `toPOJO` method or when explicitly specifying `serialize(entity, { populate: ['priority'] })`.

At the same time, in Identity Map, this (priority) is the same class.

After returning a [line of code](https://github.com/mikro-orm/mikro-orm/commit/3850c1e2b2581ea205c39e80171acb039f10561d), the test case is working correctly.

### Reproduction

https://github.com/mrmlnc/mikro-orm-qb-relations/blob/2693b8190d8eb07a1172f4f8f7db06644c542e98/src/example.test.ts#L121-L125

### What driver are you using?

@mikro-orm/sqlite

### MikroORM version

next

### Node.js version

20.18.1

### Operating system

macOS

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Discord](https://discord.gg/w8bjxFHS7X).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by B4nan** [comment]
Sounds like you were relying on a buggy behavior, your test case shouldn't behave this way, relations need to be explicitly populated, and that works based on paths. You can't just load all the target entities. `toPOJO` ignores populate hints completely, so that's why it "works". QB serialization should behave exactly the same as using `em.find`, and that line you highlighted was fixing that.

**Reply by B4nan** [reply]
Hmm way a sec, but the relation is eager, I guess I know what's happening. I got confused with the `Promise.all` and a separate `testCasePriorityRepository.findAll()`, you shouldn't need that there (and that line you highlighted might have broken this, but it shouldn't be relevant).

**Reply by B4nan** [reply]
But still, that's exactly the reason why it doesn't work, its eager, but that is not enough for QB usage, with QB you need to be explicit:

```ts
const [tcr] = await testCaseRevisionRepository.createQueryBuilder('t0')
      .select('*')
      .innerJoinAndSelect('t0.testCase', 't1', {
        version: sql`t0.version`,
      })
      .innerJoinAndSelect('t0.priority', 'p0') // added this
      .innerJoinAndSelect('t1.priority', 't2')
      .getResultList();
```
